test/test_type_promotion.py

```diff
@@ -0,0 +1,225 @@
+        torch.add(f, f, out=d)
+        self.assertEqual(d.dtype, torch.double)
+        self.assertEqual(f + f, d)
+
+    def test_mixed_type_backward(self):
+        f = torch.ones([3, 3], dtype=torch.float, requires_grad=True, device=self.device)
+        ten = torch.tensor([10], dtype=torch.double, device=self.device)
+        tens = f * ten
+        s = (tens + 2).sum()
+        s.backward()
+        self.assertEqual(f.grad, tens)
+
+        # If we don't convert the returned grad_input to the actual input type
+        # we get an error like:
+        # RuntimeError: Function SubBackward0 returned an invalid gradient at index 0 - expected type \
+        # torch.FloatTensor but got torch.DoubleTensor
+        f_dtypes = [torch.float, torch.double]
+        f_dtypes = f_dtypes if self.device != 'cuda' else f_dtypes + [torch.half]
+        i_dtypes = [torch.int, torch.long]
+        for f in ['add', 'sub', 'rsub', 'mul', 'div']:
+            for dtype1 in f_dtypes:
+                for dtype2 in (f_dtypes + i_dtypes):
+                    x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
+                    y = torch.ones(10, dtype=dtype2, device=self.device)
+
+                    func = getattr(torch, f)
+                    func(x, y).sum().backward()
+
+    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
+    # where that should hold.
+    def test_many_promotions(self):
+        from_to = {
+            torch.float16: torch.float32,
+            torch.half: torch.float16,
+            torch.int: torch.long,
+            torch.uint8: torch.long,
+            torch.uint8: torch.float,
+            torch.int: torch.float,
+            torch.int: torch.double,
+            torch.int16: torch.long,
+            torch.float16: torch.double,
+            torch.bool: torch.long,
+            torch.bool: torch.float
+        }
+
+        for k, v in from_to.items():
+            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
+            b = torch.rand([3, 3], device=self.device).to(v)
+            c = a.add(b)
+            d = a.to(v).add(b)
+            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
+            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+
+    def test_non_promoting_ops(self):
+        x = torch.ones(4, dtype=torch.double)
+        err = 'expected dtype .ouble .*but got dtype .loat'
+        self.assertRaisesRegex(RuntimeError, err,
+                               lambda: torch.neg(torch.ones(4, dtype=torch.float), out=x))
+        self.assertRaisesRegex(RuntimeError, err,
+                               lambda: torch.lerp(x, torch.ones(4, dtype=torch.float), 1))
+
+    def test_alpha_mismatch(self):
+        x = torch.ones(4, dtype=torch.int)
+        err = 'alpha must not be'
+        self.assertRaisesRegex(RuntimeError, err,
```

Comment:
that test looks good!  I meant, in addition, a valid test where alpha is a bool, assuming that is a valid invocation of the function..

Reply:
we had decided not to support bool alpha. I'll make that more explicit though so the error message makes more sense. 

Right now `torch.add(1, 1, alpha=True)` gives `For integral input tensors, argument alpha must not be a floating point number.` because we had opted to use `alpha.isIntegral(false)`.

Reply:
when did we decide that?

Reply:
ahh, I was thinking of this comment of yours on `norm_kernel_cuda_impl`:
https://github.com/pytorch/pytorch/pull/22273/commits/4d412c6dc644324bd5a46eb5aaadb7a6bc9f4ee9#r317707635

I'll change it back to accept bools.

Reply:
fixed